File: core/src/delegate.rs
//! Delegation (Phase E): the CPA hands a task to a relay-addressable peer
//! agent instead of calling a runner directly.
//!
//! Wire: NIP-29-style channel messages (kind 9) in a CPA-created OPEN
//! channel (kind 9007 create, `visibility=public` — any member may post;
//! the BUZZ private-channel roster gap is sidestepped, and 9007/9 pass the
//! stock ingest — no patch gate). A request and its result are correlated
//! by an opaque `id` echoed in the content envelopes:
//! `{"type":"job-request","id":..,"task":..}` -> peer ->
//! `{"type":"job-result","id":..,"ok":..,"out":..}`.
//! The PEER executes the task via its OWN runner-direct path (the delegation
//! mode transition: CPA -> peer -> runner, versus CPA -> runner).

use serde_json::{Value, json};

pub const CHANNEL_CREATE_KIND: u32 = 9007;
pub const STREAM_MSG_KIND: u32 = 9;

/// The delegation envelopes.
pub const JOB_REQUEST: &str = "job-request";
pub const JOB_RESULT: &str = "job-result";

/// A parsed delegation envelope (request or result).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Envelope {
    pub ty: String,
    pub id: String,
    pub task: Option<String>,
    pub ok: Option<bool>,
    pub out: Option<String>,
}

pub fn request_content(id: &str, task: &str) -> String {
    json!({ "type": JOB_REQUEST, "id": id, "task": task }).to_string()
}

pub fn result_content(id: &str, ok: bool, out: &str) -> String {
    json!({ "type": JOB_RESULT, "id": id, "ok": ok, "out": out }).to_string()
}

pub fn parse_envelope(content: &str) -> Option<Envelope> {
    let v: Value = serde_json::from_str(content).ok()?;
    Some(Envelope {
        ty: v.get("type")?.as_str()?.to_string(),
        id: v.get("id")?.as_str()?.to_string(),
        task: v.get("task").and_then(Value::as_str).map(String::from),
        ok: v.get("ok").and_then(Value::as_bool),
        out: v.get("out").and_then(Value::as_str).map(String::from),
    })
}

/// Create (idempotently) the auto-ops OPEN channel: kind-9007 with
/// `h`/`name`/`visibility=public`. Re-publishing with the same h is a no-op
/// for our purposes (creator == the CPA, auto-member).
pub fn ensure_channel(
    relay_url: &str,
    secret: &[u8; 32],
    channel_id: &str,
    name: &str,
) -> Result<(), String> {
    let event = signed_event_json(
        secret,
        CHANNEL_CREATE_KIND,
        vec![
            vec!["h".into(), channel_id.into()],
            vec!["name".into(), name.into()],
            vec!["visibility".into(), "open".into()],
        ],
        "",
    )?;
    crate::relay_http::publish_event_json(relay_url, secret, &event)
}

/// Post a channel message (kind 9), p-tagging the counterparty.
pub fn post_message(
    relay_url: &str,
    secret: &[u8; 32],
    channel_id: &str,
    mention_pubkey: &str,
    content: &str,
) -> Result<(), String> {
    let event = signed_event_json(
        secret,
        STREAM_MSG_KIND,
        vec![
            vec!["h".into(), channel_id.into()],
            vec!["p".into(), mention_pubkey.into()],
        ],
        content,
    )?;
    crate::relay_http::publish_event_json(relay_url, secret, &event)
}

/// Poll the channel for messages newer than `since`, NO p-filter (the real
/// relay serves an unfiltered kind-9 query fine, but a #p-FILTERED kind-9
/// query HANGS for some identities — verified live, so requester + executor
/// filter by author/id client-side). Every message is locally
/// signature-verified; returns (created_at, content, author) sorted.
pub fn poll_stream(
    relay_url: &str,
    secret: &[u8; 32],
    channel_id: &str,
    since: i64,
) -> Result<Vec<(i64, String, String)>, String> {
    let events = crate::relay_http::query_events(
        relay_url,
        secret,
        json!([{
            "kinds": [STREAM_MSG_KIND],
            "#h": [channel_id],
            "since": since,
            "limit": 100,
        }]),
    )?;
    collect_verified(events)
}

/// Local signature verify + sort for a poll result set (unfiltered and
/// p-filtered polls share it).
fn collect_verified(events: Vec<Value>) -> Result<Vec<(i64, String, String)>, String> {
    let mut out: Vec<(i64, String, String)> = Vec::new();
    for ev in events {
        let author = ev["pubkey"].as_str().unwrap_or("");
        let created_at = ev["created_at"].as_i64().unwrap_or(0);
        // Local signature verify over the event fields (the reader trusts
        // only genuinely-signed messages, whatever the bridge did).
        let tags: Vec<Vec<String>> = ev["tags"]
            .as_array()
            .map(|t| {
                t.iter()
                    .map(|a| {
                        a.as_array()
                            .map(|x| {
                                x.iter()
                                    .filter_map(Value::as_str)
                                    .map(String::from)
                                    .collect()
                            })
                            .unwrap_or_default()
                    })
                    .collect()
            })
            .unwrap_or_default();
        let content = ev["content"].as_str().unwrap_or("");
        if crate::nip98::verify_event(
            author,
            created_at,
            STREAM_MSG_KIND,
            &tags,
            content,
            ev["sig"].as_str().unwrap_or(""),
        )
        .is_err()
        {
            continue;
        }
        out.push((created_at, content.to_string(), author.to_string()));
    }
    out.sort_by_key(|(ts, _, _)| *ts);
    Ok(out)
}

/// Poll the channel for messages by `mention_pubkey` (p-tag) newer than
/// `since` — the p-FILTERED kind-9 query WORKS for the executor identity on
/// the live relay (it hung only for the long-lived requester identity); the
/// requester uses [`poll_stream`] instead. Locally signature-verified;
/// returns (created_at, content, author) sorted.
pub fn poll_stream_p(
    relay_url: &str,
    secret: &[u8; 32],
    channel_id: &str,
    mention_pubkey: &str,
    since: i64,
) -> Result<Vec<(i64, String, String)>, String> {
    let events = crate::relay_http::query_events(
        relay_url,
        secret,
        json!([{
            "kinds": [STREAM_MSG_KIND],
            "#h": [channel_id],
            "#p": [mention_pubkey],
            "since": since,
            "limit": 100,
        }]),
    )?;
    collect_verified(events)
}

/// Build a signed event JSON object (shared by the channel helpers).
fn signed_event_json(
    secret: &[u8; 32],
    kind: u32,
    tags: Vec<Vec<String>>,
    content: &str,
) -> Result<String, String> {
    let ts = crate::auth::now_secs();
    let (pubkey, id, sig) = crate::nip98::sign_event(secret, kind, ts, tags.clone(), content)
        .map_err(|e| e.to_string())?;
    Ok(json!({
        "id": id,
        "pubkey": pubkey,
        "created_at": ts,
        "kind": kind,
        "tags": tags,
        "content": content,
        "sig": sig,
    })
    .to_string())
}

File: contract/delegate/delegate.go
// Package delegate reproduces core/src/delegate.rs — the Phase-E delegation
// wire: CPA hands a task to a relay-addressable peer agent via kind-9 channel
// messages in a CPA-created OPEN channel. Requests/results correlate by an
// echoed `id` in the content envelopes.
package delegate

import (
	"encoding/json"
	"sort"
	"strconv"
	"time"

	"freehold/contract/relay"
	"freehold/contract/wire"
)

const (
	ChannelCreateKind = wire.ChannelCreate
	StreamMsgKind     = wire.ChannelMessage
	JobRequest        = "job-request"
	JobResult         = "job-result"
)

// Envelope is a parsed delegation envelope (request or result).
type Envelope struct {
	Type string  `json:"type"`
	ID   string  `json:"id"`
	Task *string `json:"task,omitempty"`
	OK   *bool   `json:"ok,omitempty"`
	Out  *string `json:"out,omitempty"`
}

// RequestContent builds a job-request envelope JSON.
func RequestContent(id, task string) string {
	b, _ := json.Marshal(map[string]interface{}{"type": JobRequest, "id": id, "task": task})
	return string(b)
}

// ResultContent builds a job-result envelope JSON.
func ResultContent(id string, ok bool, out string) string {
	b, _ := json.Marshal(map[string]interface{}{"type": JobResult, "id": id, "ok": ok, "out": out})
	return string(b)
}

// ParseEnvelope parses a delegation envelope.
func ParseEnvelope(content string) *Envelope {
	var e Envelope
	if err := json.Unmarshal([]byte(content), &e); err != nil {
		return nil
	}
	if e.Type == "" || e.ID == "" {
		return nil
	}
	return &e
}

// EnsureChannel creates (idempotently) the auto-ops OPEN channel (kind 9007,
// h/name/visibility=open).
func EnsureChannel(relayURL string, secret []byte, channelID, name string) error {
	return EnsureChannelAuth(relayURL, relayURL, secret, channelID, name)
}

// EnsureChannelAuth is EnsureChannel with a separate NIP-98 auth URL (the
// pre-Caddy LAN-dial case in agent-tools).
func EnsureChannelAuth(dialURL, authURL string, secret []byte, channelID, name string) error {
	return publishSignedAuth(dialURL, authURL, secret, ChannelCreateKind, [][]string{
		{"h", channelID},
		{"name", name},
		{"visibility", "open"},
	}, "")
}

// PostMessage posts a kind-9 channel message, p-tagging the counterparty.
func PostMessage(relayURL string, secret []byte, channelID, mentionPubkey, content string) error {
	return publishSigned(relayURL, secret, StreamMsgKind, [][]string{
		{"h", channelID},
		{"p", mentionPubkey},
	}, content)
}

// PollResult is one verified stream message (created_at, content, author).
type PollResult struct {
	CreatedAt int64
	Content   string
	Author    string
}

// PollStream polls the channel for messages newer than since, NO p-filter,
// locally signature-verified, sorted. Reproduces delegate::poll_stream (the
// requester uses the unfiltered path — a #p-filtered query hangs for some
// identities, verified live).
func PollStream(relayURL string, secret []byte, channelID string, since int64) ([]PollResult, error) {
	return poll(relayURL, secret, channelID, "", since)
}

// PollStreamP polls for messages by mention_pubkey (p-tag) newer than since
// (the executor uses the p-filtered path, which works for its identity).
func PollStreamP(relayURL string, secret []byte, channelID, mentionPubkey string, since int64) ([]PollResult, error) {
	return poll(relayURL, secret, channelID, mentionPubkey, since)
}

func publishSigned(relayURL string, secret []byte, kind uint32, tags [][]string, content string) error {
	return publishSignedAuth(relayURL, relayURL, secret, kind, tags, content)
}

func publishSignedAuth(dialURL, authURL string, secret []byte, kind uint32, tags [][]string, content string) error {
	ts := time.Now().Unix()
	pubkey, id, sig, err := wire.SignEvent(secret, kind, ts, tags, content)
	if err != nil {
		return err
	}
	ev := map[string]interface{}{
		"id": id, "pubkey": pubkey, "created_at": ts, "kind": kind,
		"tags": tags, "content": content, "sig": sig,
	}
	evBytes, _ := json.Marshal(ev)
	return relay.PublishEventJSONAuth(dialURL, authURL, secret, string(evBytes))
}

func poll(relayURL string, secret []byte, channelID, p string, since int64) ([]PollResult, error) {
	filter := map[string]interface{}{
		"kinds": []interface{}{StreamMsgKind},
		"#h":    []interface{}{channelID},
		"since": since,
		"limit": 100,
	}
	if p != "" {
		filter["#p"] = []interface{}{p}
	}
	events, err := relay.QueryEvents(relayURL, secret, []interface{}{filter})
	if err != nil {
		return nil, err
	}
	out := collectVerified(events)
	sort.Slice(out, func(i, j int) bool { return out[i].CreatedAt < out[j].CreatedAt })
	return out, nil
}

// collectVerified locally signature-verifies every event.
func collectVerified(events []map[string]interface{}) []PollResult {
	var out []PollResult
	for _, ev := range events {
		author, _ := ev["pubkey"].(string)
		createdAt := int64num(ev["created_at"])
		tags := parseTags(ev)
		content, _ := ev["content"].(string)
		sig, _ := ev["sig"].(string)
		if _, err := wire.VerifyEvent(author, createdAt, StreamMsgKind, tags, content, sig); err != nil {
			continue
		}
		out = append(out, PollResult{CreatedAt: createdAt, Content: content, Author: author})
	}
	return out
}

func parseTags(ev map[string]interface{}) [][]string {
	var tags [][]string
	if raw, ok := ev["tags"].([]interface{}); ok {
		for _, t := range raw {
			var row []string
			if arr, ok := t.([]interface{}); ok {
				for _, a := range arr {
					if s, ok := a.(string); ok {
						row = append(row, s)
					}
				}
			}
			tags = append(tags, row)
		}
	}
	return tags
}

func int64num(v interface{}) int64 {
	switch n := v.(type) {
	case float64:
		return int64(n)
	case int64:
		return n
	case json.Number:
		i, _ := n.Int64()
		return i
	case string:
		i, _ := strconv.ParseInt(n, 10, 64)
		return i
	}
	return 0
}
